### ISSUE TITLE
[Bug]: Expert parallelism drops every token routed to a non-local expert

### ISSUE THREAD

**Issue by mantreshkhurana** [Spyxpo/framerai#230]
### Preflight checklist

- [x] I searched existing issues and did not find a duplicate.
- [x] I am running the latest version of the affected branch.

### Affected component

Model (Python)

### Description

`shard_experts` gives each rank its slice of the experts and records the layout on the module:
`model/training/expert_parallel.py:89-90` sets `moe.ep_plan = plan` and `moe.expert_offset = start`.

`MoEFeedForward.forward` (`model/modules/moe.py:64-85`) never reads `ep_plan`. The only other
mention of the attribute in the whole tree is its initialisation to `None` at
`model/modules/moe.py:54`, and `all_to_all_dispatch` and `all_to_all_combine`
(`model/training/expert_parallel.py:112`, `model/training/expert_parallel.py:134`) have no call
sites anywhere.

The dispatch path is written for the local slice and stops there.
`_grouped_expert_forward` filters the routed assignments to `[expert_offset, expert_offset + E_local)`
at `model/modules/moe.py:113-116`, returns `torch.zeros_like(x_flat)` when nothing is local
(`model/modules/moe.py:120-121`), and scatters only the local contributions with `index_add_`
at `model/modules/moe.py:161-162`. Nothing sends the remaining tokens to the rank that owns their
expert, and nothing all-reduces the result over the expert-parallel group.

So with `ep_world=8`, each rank's MoE output is missing roughly seven eighths of the expert
contributions the router selected, and each rank has a different eighth. The model trains and
serves on a silently truncated feed-forward. Nothing raises, and the loss curve still descends,
which is what makes it costly: an expert-parallel run looks like it is working.

### Steps to reproduce

1. Construct a small MoE model and run a forward pass on one rank with no sharding; record the
   output.
2. Launch the same model under `torchrun --nproc_per_node=2` and shard the experts with
   `shard_experts` over a 2-way expert-parallel mesh.
3. Run the same input through the sharded model.
4. Compare the outputs: they differ, and neither rank matches the unsharded reference.
5. Confirm the cause by checking that `moe.ep_plan` is set and that no all-to-all or all-reduce
   is executed during the forward pass.

### Expected behavior

An expert-parallel forward should be numerically equivalent to the unsharded forward, within
floating-point tolerance. The routed tokens should be sent to the rank owning their expert,
processed there, and returned, using the `all_to_all_dispatch` and `all_to_all_combine`
primitives that already exist for the purpose.

Until that is wired, a module constructed with `expert_offset != 0` while `ep_plan` is unused
should fail loudly at construction rather than compute a partial result.

### Logs and error output

```shell
# 2-way expert parallelism, same input, same seed
reference (unsharded)  out[0, 0, :4] = [ 0.4127, -0.2213,  0.0885,  0.3390]
rank 0     (ep_world=2) out[0, 0, :4] = [ 0.2011, -0.0043,  0.0885,  0.1170]
rank 1     (ep_world=2) out[0, 0, :4] = [ 0.2116, -0.2170,  0.0000,  0.2220]

$ grep -rn "all_to_all_dispatch\|all_to_all_combine" model/modules/
(no matches)
```

### Environment

OS: Ubuntu 22.04
Python: 3.12.4
PyTorch: 2.8.0
Node: 22.11.0
GPU/VRAM: 8x A100 80GB
Affected branch: stable

**Comment by kunaldeo001** [comment]
I’d like to work on Issue
